File: PA_17.py
#LC 525 : Contiguous Array
#Time Complexity - O(n)
#Space Complexity - O(n)
d ={0:-1}
ml,c = 0,0
if nums.count(0) == nums.count(1):
    return len(nums)
for i in range(len(nums)):
    if nums[i] == 1:
        c= c + 1
    else:
        c = c - 1
    if c in d.keys():
        if ml < i - d[c]:
            ml = i - d[c]
    else:
        d[c] = i

return ml

File: PA17.java
//Leetcode 525. Contiguous Array 
//Time complexity : O(n) // n is length of nums array
//Space Complexity : O(n) // / n is length of nums array

class Solution {
    public int findMaxLength(int[] nums) {
       int ones=0;
        int count=0;
        if(nums==null || nums.length==0){
            return 0;
        }
        HashMap<Integer,Integer> hm = new HashMap<>();
        hm.put(0,-1);
        int temp=0;
        for(int i=0;i<nums.length;i++){
            if(nums[i]==1){
                ones=ones+1;
            }
            else{
                ones=ones-1;
            }
            if(hm.containsKey(ones)){
                 temp=i-hm.get(ones);
                if(temp>count){
                    count=temp;
                }
            }
            else{
                hm.put(ones,i);
            }
            
           
        }
        return count;
    }
}